### ISSUE TITLE
Refactor NTXentLoss

### ISSUE THREAD

**Issue by philippmwirth** [lightly-ai/lightly#268]
# Refactor NTXentLoss

The implementation of the NTXentLoss is cumbersome and overly complicated. A simple implementation is possible using `torch.einsum`. We should refactor the code to make it easier for our users to understand what's going on.

**Comment by philippmwirth** [comment]
@IgorSusmelj adding this would mean to remove the version of the loss which uses squared distance instead of cosine similarity. Are you ok with that?

**Comment by MalteEbner** [comment]
> @IgorSusmelj adding this would mean to remove the version of the loss which uses squared distance instead of cosine similarity. Are you ok with that?

There is no L2 distance in the current implementation anyway, only the dot similarity. But as the dot similarity and cosine similarity of two normalised vectors are the same, it essentially boils down to only cosine similarity.

**Comment by philippmwirth** [comment]
[This](https://github.com/philippmwirth/co2/blob/main/co2.py) could serve as inspiration

**Comment by MalteEbner** [comment]
What I do not really get from the papers is what servers as negatives examples if there is not memory bank. From the SIMCLR paper, I get the following impression
```
loss = 
loss(out0_positive, out0_negatives) + loss(out0_positive, out1_negatives) 
+ loss(out1_positives, out0_negatives) + loss(out1_positive, out1_negative)
= loss_0_0 + loss_0_1 + loss_1_0 + loss_1_1
```
However, in the CO2 example, only loss_0_1 is taken as loss. Am I correct, that the NTXent loss uses all 4 losses as above?

**Comment by philippmwirth** [comment]
I think you are right. The current implementation also does it like this:

https://github.com/lightly-ai/lightly/blob/239123cec1031199a53c9863b5f806b5091927b1/lightly/loss/ntx_ent_loss.py#L147

Can you quickly confirm that you understand the current code the same way I do?
